Allow to disable attribute discovery in ActiveRecord

kirs opened the pull request:
This works as an opposite to `ignore_columns`. Sometimes we want to explicitly declare used columns with Attributes API in the model and use only them, ignoring the rest of columns:

```ruby
class Cat < ActiveRecord::Base
  self.attribute_auto_discovery = false

  attribute :name, :string
  attribute :engines_count, :integer
  attribute :created_at, :datetime
  attribute :updated_at, :datetime
end
```

@wvanbergen @rafaelfranca @matthewd

kirs committed:
Allow to disable attribute discovery in ActiveRecord

Sometimes you want to explicitly declare attributes in the model
with Attributes API and ignore the rest of table columns.

simi commented on activerecord/lib/active_record/model_schema.rb:
```diff
@@ -235,6 +235,29 @@ def table_name=(value)
+      #
+      # Despite on the fact that `projects` table has more columns that `name` and `members_count`,
+      # ActiveRecord will only use explicitly declared attributes
+      def attribute_auto_discovery
```
Can't be this moved to `included` block and defined via `mattr_accessor` or `class_attribute` with default value as another options?

matthewd commented on activerecord/lib/active_record/model_schema.rb:
```diff
@@ -448,6 +471,11 @@ def load_schema
           @columns_hash = connection.schema_cache.columns_hash(table_name).except(*ignored_columns)
+
+          unless attribute_auto_discovery
+            @columns_hash.reject! { |k| !attributes_to_define_after_schema_loads.key?(k) }
```
Why is this filtering necessary? Why can't we just skip the schema load entirely? :confused:

sgrif replied:
Agreed. This should definitely just be `return unless attribute_auto_discovery`

kirs replied:
This filtering in necessary because otherwise `@columns_hash` will include all table columns. It's loaded in the line above.

We already do this filter for ignored columns (`.except(*ignored_columns)`)

sgrif commented:
I'd prefer something like `define_attributes_from_schema` for the option name.

kaspth commented:
Generally if the default is true for a new configuration is true, we invert the name: `self.skip_defining_attributes_from_schema = true`.

See https://github.com/rails/rails/pull/26976#issuecomment-261801424 for more.

wvanbergen commented:
- `id` is automatically defined?
- Should you define `foo_id`, or should `belongs_to :foo` take care of that?

sgrif commented:
> id is automatically defined?

`id` should continue to refer to the primary key. If `self.primary_key` is set to something other than `id`, it will need to be defined.

> Should you define foo_id, or should belongs_to :foo take care of that?

`foo_id` should need to be manually defined. The only reason `id` is treated specially is because the method `id` exists without the column

sgrif commented:
Note: I want to be conservative to start. We can always change it to have `foo_id` defined automatically later without breaking backwards compat.

kirs commented:
Thanks everyone for the feedback!
I've updated the PR and it's ready for another round of 👀

sgrif commented on activerecord/lib/active_record/model_schema.rb:
```diff
@@ -149,6 +176,9 @@ module ModelSchema
       self.ignored_columns = [].freeze
 
+      class_attribute :skip_defining_attributes_from_schema, instance_accessor: false
+      self.skip_defining_attributes_from_schema = false
```
Can we drop the `skip_` and have it default to `true`?

sgrif commented on activerecord/lib/active_record/model_schema.rb:
```diff
@@ -447,7 +477,30 @@ def load_schema
+
+          if skip_defining_attributes_from_schema
+            if ignored_columns.any?
+              raise ArgumentError, "can't use ignored_columns and skip_defining_attributes_from_schema at the same time"
```
Let's handle this in the writer method for `define_attributes_from_schema` and `ignored_columns`. We want to error early, and we shouldn't assume this is the only place the two can interact.

sgrif commented on activerecord/lib/active_record/model_schema.rb:
```diff
@@ -447,7 +477,30 @@ def load_schema
+              raise ArgumentError, "can't use ignored_columns and skip_defining_attributes_from_schema at the same time"
+            end
+
+            if primary_key && !attributes_to_define_after_schema_loads.key?(primary_key)
```
This feels like it should be unnecessary.

kirs commented:
I've addressed all concerns 👌 

> Let's handle this in the writer method for define_attributes_from_schema and ignored_columns. We want to error early, and we shouldn't assume this is the only place the two can interact.

For that, I had to move away from `class_attribute` and define custom accessors.

kirs commented:
🙏 👀 when you have a minute.

sgrif commented on activerecord/lib/active_record/model_schema.rb:
```diff
@@ -447,7 +500,14 @@ def load_schema
-          @columns_hash = connection.schema_cache.columns_hash(table_name).except(*ignored_columns)
+          @columns_hash = connection.schema_cache.columns_hash(table_name)
+
+          unless define_attributes_from_schema
```
This seems like the wrong place to handle things. We should not be modifying `columns_hash` based on this setting (other than perhaps not setting it at all). It seems like we should be skipping the loop down below instead.

kirs commented:
@sgrif I'm ready to invest time into finishing this feature and addressing the feedback, but is it something that you'd like to see in ActiveRecord?

robotdana commented:
is this something that is being worked on? I would like to use a feature like this

rafaelfranca commented:
@robotdana yes! I was planning to have someone in my team to work on it but we are happy to also review a PR if it is open.

robotdana commented:
https://github.com/rails/rails/pull/35309 I have a PR ready, rebasing and finishing this. Unfortunately @sgrif it looked like not modifying the columns_hash was more complex than not, and if we changed it we should also change the existing implementation of ignored_columns (which does modify the columns_hash)

kirs commented:
Closing in favour of https://github.com/rails/rails/pull/35309
